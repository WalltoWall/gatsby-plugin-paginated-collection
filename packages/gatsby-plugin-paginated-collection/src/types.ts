import {
  PluginOptions as GatsbyPluginOptions,
  NodeInput,
  CreatePagesArgs,
  Node,
} from 'gatsby'

export type ID = string

export enum NodeType {
  Collection = 'PaginatedCollection',
  Page = 'PaginatedCollectionPage',
}

export type NormalizedNode = Record<string, unknown>

export type PageNode = Node & PageNodeInput

export interface PageNodeInput extends NodeInput {
  collection: ID
  index: number
  nextPage: ID | undefined
  hasNextPage: boolean
  previousPage: ID | undefined
  hasPreviousPage: boolean
  nodeCount: number
  nodes: NormalizedNode[]
  internal: NodeInput['internal'] & {
    type: NodeType.Page
  }
}

export type CollectionNode = Node & CollectionNodeInput

export interface CollectionNodeInput extends NodeInput {
  name: string
  pageSize: number
  firstPageSize: number
  lastPageSize: number
  nodeCount: number
  pageCount: number
  pages: ID[]
  internal: NodeInput['internal'] & {
    type: NodeType.Collection
  }
}

export interface GraphQLResult {
  errors?: any
  data?: unknown
}

export type PluginOptions = Required<ProvidedPluginOptions>

export interface PluginConfig {
  resolve: string
  options?: Record<string, unknown>
}

export interface ProvidedPluginOptions extends GatsbyPluginOptions {
  name: string
  query: string
  pageSize?: number
  firstPageSize?: number
  normalizer: (input: GraphQLResult) => NormalizedNode[]
  plugins: PluginConfig[]
}

export interface Plugin {
  onPostCreateNodes?(
    gatsbyContext: CreatePagesArgs & { traceId: 'initial-createPages' },
    rootPluginOptions: PluginOptions,
    pluginOptions: unknown,
  ): void | Promise<void>
}
