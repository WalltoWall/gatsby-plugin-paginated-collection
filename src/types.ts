import { PluginOptions as GatsbyPluginOptions, NodeInput } from 'gatsby'

export type ID = string

export enum NodeType {
  Collection = 'PaginatedCollection',
  Page = 'PaginatedCollectionPage',
}

export interface PageNodeInput extends NodeInput {
  collection: ID
  index: number
  nextPage: ID | undefined
  hasNextPage: boolean
  previousPage: ID | undefined
  hasPreviousPage: boolean
  nodeCount: number
  nodes: Node[]
  internal: NodeInput['internal'] & {
    type: NodeType.Page
  }
}

export interface CollectionNodeInput extends NodeInput {
  name: string
  pageSize: number
  firstPageSize: number
  lastPageSize: number
  nodeCount: number
  pageCount: number
  pages: ID[]
  internal: NodeInput['internal'] & {
    type: NodeType.Collection
  }
}

export interface GraphQLResult {
  errors?: any
  data?: unknown
}

export type Node = Record<string, unknown>

export type PluginOptions = Required<ProvidedPluginOptions>

export interface ProvidedPluginOptions extends GatsbyPluginOptions {
  name: string
  query: string
  pageSize?: number
  firstPageSize?: number
  normalizer: (input: GraphQLResult) => Node[]
}
