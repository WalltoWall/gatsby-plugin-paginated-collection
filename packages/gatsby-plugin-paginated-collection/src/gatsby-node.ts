import {
  GatsbyNode,
  CreateSchemaCustomizationArgs,
  PluginOptions as GatsbyPluginOptions,
} from 'gatsby'

import { chunk, fmtMsg, createCollectionNode } from './utils'
import { ProvidedPluginOptions, PluginOptions, Plugin } from './types'
import { types } from './gqlTypes'

export const DEFAULT_PLUGIN_OPTIONS = {
  pageSize: 10,
}

export const createPages: GatsbyNode['createPages'] = (
  gatsbyContext,
  providedPluginOptions: ProvidedPluginOptions,
  callback,
) => {
  const pluginOptions: PluginOptions = {
    ...DEFAULT_PLUGIN_OPTIONS,
    ...providedPluginOptions,
    firstPageSize:
      providedPluginOptions.firstPageSize ??
      providedPluginOptions.pageSize ??
      DEFAULT_PLUGIN_OPTIONS.pageSize,
  }

  const { graphql, getNode, reporter } = gatsbyContext
  const { name, query, pageSize, normalizer, plugins } = pluginOptions

  const asyncFn = async () => {
    const queryResult = await graphql(query)

    if (queryResult.errors) {
      reporter.error(
        fmtMsg(
          `The provided GraphQL query for "${name}" contains errors. Pagination files will not be created.`,
        ),
        queryResult.errors[0],
      )
      return
    }

    const items = await Promise.resolve(normalizer(queryResult))

    if (!Array.isArray(items)) {
      reporter.error(
        fmtMsg(
          `The normalizer function for "${name}" did not return an array. Pagination files will not be created.`,
        ),
      )
      return
    }

    if (items.length < 1)
      reporter.warn(
        fmtMsg(
          `No items for "${name}" were returned. Pagination data will be created on an empty set.`,
        ),
      )

    const chunks = chunk(pageSize, items)
    const nodeId = createCollectionNode(chunks, gatsbyContext, pluginOptions)
    const node = getNode(nodeId)

    for (const plugin of plugins) {
      const requiredPlugin: Plugin = require(plugin.resolve)
      await requiredPlugin?.onPostCreateNodes?.(
        node,
        plugin.options,
        gatsbyContext,
        pluginOptions,
      )
    }
  }

  asyncFn().finally(() => callback && callback(null))
}

export const createSchemaCustomization: GatsbyNode['createSchemaCustomization'] = async (
  gatsbyContext: CreateSchemaCustomizationArgs,
  _pluginOptions: GatsbyPluginOptions,
) => {
  const { actions } = gatsbyContext
  const { createTypes } = actions

  createTypes(types)
}
