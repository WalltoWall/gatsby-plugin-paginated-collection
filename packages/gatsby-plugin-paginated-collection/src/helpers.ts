import { CreatePagesArgs } from 'gatsby'

import {
  NormalizedNode,
  NodeType,
  ID,
  PageNodeInput,
  CollectionNodeInput,
} from './types'
import { chunk } from './utils'

interface CreatePageNodeArgs {
  /** Nodes for the page. */
  chunk: NormalizedNode[]
  /** UUID for the page. Used as a pseudo cursor to identify the page. */
  id: ID
  /** Index of the page within the collection. */
  index: number
  /** Tuples Tuples of page UUIDs and nodes. */
  chunkTuples: [ID, NormalizedNode[]][]
  collectionId: ID
  /** Gatsby createPages context. */
  createNode: CreatePagesArgs['actions']['createNode']
  /** Options provided to the plugin by the user. */
  createContentDigest: CreatePagesArgs['createContentDigest']
}

/**
 * Creates a Gatsby Node representing a page in a collection of nodes.
 *
 * @param args Arguments to create the page node.
 *
 * @returns The created Gatsby Node.
 */
const createPageNode = (args: CreatePageNodeArgs): PageNodeInput => {
  const {
    chunk,
    id,
    index,
    chunkTuples,
    collectionId,
    createNode,
    createContentDigest,
  } = args

  const nextChunkTuple = chunkTuples[index + 1]
  const previousChunkTuple = chunkTuples[index - 1]

  const node: PageNodeInput = {
    id,
    collection: collectionId,
    index,
    nextPage: nextChunkTuple?.[0],
    hasNextPage: Boolean(nextChunkTuple),
    previousPage: previousChunkTuple?.[0],
    hasPreviousPage: Boolean(previousChunkTuple),
    nodeCount: chunk.length,
    nodes: chunk,
    internal: {
      type: NodeType.Page,
      contentDigest: createContentDigest(chunk),
    },
  }

  createNode(node, { name: 'gatsby-plugin-paginated-collection' })

  return node
}

interface CreatePaginatedCollectionNodesArgs {
  /** Array of items to paginate. */
  collection: NormalizedNode[]
  /** Name of the collection. */
  name: string
  /** The number of nodes to include in each page. */
  pageSize: number
  /** The action used to create nodes. */
  createNode: CreatePagesArgs['actions']['createNode']
  /** A helper function for creating node IDs. */
  createNodeId: CreatePagesArgs['createNodeId']
  /** A helper function for creating node content digests. */
  createContentDigest: CreatePagesArgs['createContentDigest']
}

/**
 * Creates a Gatsby Node and its child nodes representing a collection of pages
 * of nodes.
 *
 * @param args Arguments to create the collection nodes.
 *
 * @returns The created Gatsby Node.
 */
export const createPaginatedCollectionNodes = (
  args: CreatePaginatedCollectionNodesArgs,
) => {
  const {
    collection,
    name,
    pageSize,
    createNode,
    createNodeId,
    createContentDigest,
  } = args

  const id: string = createNodeId(`${NodeType.Collection} ${name}`)

  const chunks = chunk(pageSize, collection)
  const pageNodes = chunks
    .map<[ID, NormalizedNode[]]>((chunk, index) => [
      createNodeId(`${NodeType.Page} ${name} ${index}`),
      chunk,
    ])
    .map(([pageId, chunk], index, arr) =>
      createPageNode({
        chunk,
        id: pageId,
        index,
        chunkTuples: arr,
        collectionId: id,
        createNode,
        createContentDigest,
      }),
    )

  const nodeCount = pageNodes.reduce(
    (count, pageNode) => (count += pageNode.nodes.length),
    0,
  )

  const pageNodeIds = pageNodes.map(pageNode => pageNode.id)

  const node: CollectionNodeInput = {
    id,
    name,
    pageSize,
    firstPageSize: pageNodes[0].nodes.length,
    lastPageSize: pageNodes[pageNodes.length - 1].nodes.length,
    nodeCount,
    pageCount: pageNodes.length,
    pages: pageNodeIds,
    internal: {
      type: NodeType.Collection,
      contentDigest: createContentDigest(pageNodes),
    },
  }

  createNode(node, { name: 'gatsby-plugin-paginated-collection' })

  return node
}
