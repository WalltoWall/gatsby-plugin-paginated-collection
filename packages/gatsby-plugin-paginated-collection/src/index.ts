export { Plugin, NodeType, PageNode, CollectionNode } from './types'
