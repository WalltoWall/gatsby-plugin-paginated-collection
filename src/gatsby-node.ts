import { GatsbyNode } from 'gatsby'

import { chunk, fmtMsg, createCollectionNode } from './utils'
import { ProvidedPluginOptions, PluginOptions } from './types'

const DEFAULT_PLUGIN_OPTIONS = {
  pageSize: 10,
}

export const createPages: GatsbyNode['createPages'] = (
  gatsbyContext,
  providedPluginOptions: ProvidedPluginOptions,
  callback,
) => {
  const pluginOptions: PluginOptions = {
    ...DEFAULT_PLUGIN_OPTIONS,
    ...providedPluginOptions,
    firstPageSize:
      providedPluginOptions.firstPageSize ??
      providedPluginOptions.pageSize ??
      DEFAULT_PLUGIN_OPTIONS.pageSize,
  }

  const { graphql, reporter } = gatsbyContext
  const { name, query, pageSize, normalizer } = pluginOptions

  const asyncFn = async () => {
    const queryResult = await graphql(query)

    if (queryResult.errors) {
      reporter.error(
        fmtMsg(
          `The provided GraphQL query for "${name}" contains errors. Pagination files will not be created.`,
        ),
        queryResult.errors[0],
      )
      return
    }

    const items = await Promise.resolve(normalizer(queryResult))

    if (!Array.isArray(items)) {
      reporter.error(
        fmtMsg(
          `The normalizer function for "${name}" did not return an array. Pagination files will not be created.`,
        ),
      )
      return
    }

    if (items.length < 1)
      reporter.warn(
        fmtMsg(
          `No items for "${name}" were returned. Pagination data will be created on an empty set.`,
        ),
      )

    const chunks = chunk(pageSize, items)
    createCollectionNode(chunks, gatsbyContext, pluginOptions)
  }

  asyncFn().finally(() => callback && callback(null))
}
