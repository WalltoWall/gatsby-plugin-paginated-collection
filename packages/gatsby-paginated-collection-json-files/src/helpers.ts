import path from 'path'
import { promises as fs } from 'fs'
import { CollectionNode, PageNode } from 'gatsby-plugin-paginated-collection'
import { CreatePagesArgs } from 'gatsby'

export interface WritePaginatedCollectionJSONFilesArgs {
  node: CollectionNode
  directory: string
  expand?: ('nextPage' | 'previousPage' | 'collection')[]
  filename?: 'id' | 'index' | ((node: PageNode) => string)
  getNode: CreatePagesArgs['getNode']
}

export const writePaginatedCollectionJSONFiles = async (
  args: WritePaginatedCollectionJSONFilesArgs,
): Promise<void> => {
  const { node, directory, expand = [], filename = 'id', getNode } = args

  await fs.mkdir(directory, { recursive: true })

  await Promise.all(
    node.pages.map(async (pageId) => {
      const page: PageNode = getNode(pageId)
      const trimmedPage = {
        ...page,
        nextPage: expand.includes('nextPage')
          ? {
              ...getNode(page.nextPage),
              nodes: undefined,
              internal: undefined,
              children: undefined,
              parent: undefined,
            }
          : page.nextPage,
        previousPage: expand.includes('previousPage')
          ? {
              ...getNode(page.previousPage),
              nodes: undefined,
              internal: undefined,
              children: undefined,
              parent: undefined,
            }
          : page.previousPage,
        collection: expand.includes('collection')
          ? {
              ...getNode(page.collection),
              internal: undefined,
              children: undefined,
              parent: undefined,
            }
          : page.collection,
        internal: undefined,
        children: undefined,
        parent: undefined,
      }

      const fileBasename =
        typeof filename === 'function' ? String(filename(page)) : page[filename]

      await fs.writeFile(
        path.join(directory, fileBasename + '.json'),
        JSON.stringify(trimmedPage),
      )
    }),
  )
}
