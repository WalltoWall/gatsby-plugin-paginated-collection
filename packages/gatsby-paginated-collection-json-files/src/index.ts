import path from 'path'
import { Plugin } from 'gatsby-plugin-paginated-collection'

import {
  writePaginatedCollectionJSONFiles,
  WritePaginatedCollectionJSONFilesArgs,
} from './helpers'

export { writePaginatedCollectionJSONFiles } from './helpers'

const DEFAULT_PLUGIN_OPTIONS = {
  path: 'paginated-collections',
}

export interface PluginOptions {
  path?: string
  expand?: WritePaginatedCollectionJSONFilesArgs['expand']
  filename?: WritePaginatedCollectionJSONFilesArgs['filename']
}

export const onPostCreateNodes: Plugin['onPostCreateNodes'] = async (
  node,
  rawPluginOptions: PluginOptions,
  gatsbyContext,
  _rootPluginOptions,
) => {
  const { getNode, store } = gatsbyContext
  const { program } = store.getState()
  const pluginOptions: PluginOptions &
    Required<Pick<PluginOptions, keyof typeof DEFAULT_PLUGIN_OPTIONS>> = {
    ...DEFAULT_PLUGIN_OPTIONS,
    ...rawPluginOptions,
  }

  const dir = path.resolve(program.directory, 'public', pluginOptions.path)

  await writePaginatedCollectionJSONFiles({
    node,
    directory: dir,
    expand: pluginOptions.expand,
    filename: pluginOptions.filename,
    getNode,
  })
}
