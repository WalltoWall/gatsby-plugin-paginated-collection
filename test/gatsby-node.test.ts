import { CreatePagesArgs } from 'gatsby'

import { createPages } from '../src/gatsby-node'
import { ProvidedPluginOptions } from '../src/types'

const mockActions = {
  deletePage: jest.fn(),
  createPage: jest.fn(),
  deleteNode: jest.fn(),
  deleteNodes: jest.fn(),
  createNode: jest.fn(),
  touchNode: jest.fn(),
  createNodeField: jest.fn(),
  createParentChildLink: jest.fn(),
  setWebpackConfig: jest.fn(),
  replaceWebpackConfig: jest.fn(),
  setBabelOptions: jest.fn(),
  setBabelPlugin: jest.fn(),
  setBabelPreset: jest.fn(),
  createJob: jest.fn(),
  createJobV2: jest.fn(),
  setJob: jest.fn(),
  endJob: jest.fn(),
  setPluginStatus: jest.fn(),
  createRedirect: jest.fn(),
  addThirdPartySchema: jest.fn(),
  createTypes: jest.fn(),
  createFieldExtension: jest.fn(),
}

const mockGatsbyContext: CreatePagesArgs & {
  traceId: 'initial-createPages'
} = {
  pathPrefix: 'pathPrefix',
  boundActionCreators: mockActions,
  actions: mockActions,
  loadNodeContent: jest.fn(),
  store: {
    dispatch: jest.fn(),
    subscribe: jest.fn(),
    getState: jest.fn(),
    replaceReducer: jest.fn(),
  },
  emitter: {
    addListener: jest.fn(),
    on: jest.fn(),
    once: jest.fn(),
    prependListener: jest.fn(),
    prependOnceListener: jest.fn(),
    removeListener: jest.fn(),
    off: jest.fn(),
    removeAllListeners: jest.fn(),
    setMaxListeners: jest.fn(),
    getMaxListeners: jest.fn(),
    listeners: jest.fn(),
    rawListeners: jest.fn(),
    emit: jest.fn(),
    eventNames: jest.fn(),
    listenerCount: jest.fn(),
  },
  getNodes: jest.fn(),
  getNode: jest.fn(),
  getNodesByType: jest.fn(),
  hasNodeChanged: jest.fn(),
  reporter: {
    stripIndent: jest.fn(),
    format: jest.fn(),
    setVerbose: jest.fn(),
    setNoColor: jest.fn(),
    panic: jest.fn(),
    panicOnBuild: jest.fn(),
    error: jest.fn(),
    uptime: jest.fn(),
    success: jest.fn(),
    verbose: jest.fn(),
    info: jest.fn(),
    warn: jest.fn(),
    log: jest.fn(),
    activityTimer: jest.fn(),
    createProgress: jest.fn(),
  },
  getNodeAndSavePathDependency: jest.fn(),
  cache: {
    getAndPassUp: jest.fn(),
    wrap: jest.fn(),
    set: jest.fn(),
    mset: jest.fn(),
    get: jest.fn(),
    mget: jest.fn(),
    del: jest.fn(),
    reset: jest.fn(),
  },
  createNodeId: jest.fn().mockReturnValue('createNodeId'),
  createContentDigest: jest.fn().mockReturnValue('createContentDigest'),
  tracing: {
    tracer: {},
    parentSpan: {},
    startSpan: jest.fn(),
  },
  schema: {
    buildObjectType: jest.fn(),
    buildUnionType: jest.fn(),
    buildInterfaceType: jest.fn(),
    buildInputObjectType: jest.fn(),
    buildEnumType: jest.fn(),
    buildScalarType: jest.fn(),
  },
  traceId: 'initial-createPages',
  waitForCascadingActions: false,
  parentSpan: {},
  graphql: jest.fn(),
}

const mockQueryResult = {
  data: {
    allNode: {
      edges: [{ foo: 'bar' }, { foo: 'baz' }],
    },
  },
}

const pluginOptions: ProvidedPluginOptions = {
  name: 'name',
  query: 'query',
  normalizer: queryResult =>
    (queryResult as typeof mockQueryResult).data.allNode.edges.map(node => ({
      foo: node.foo,
    })),
  plugins: [],
}

beforeAll(() => {
  ;(mockGatsbyContext.graphql as jest.Mock).mockReturnValue(
    Promise.resolve(mockQueryResult),
  )
})

describe('sourceNodes', () => {
  beforeEach(() => jest.clearAllMocks())

  test('creates nodes', async () => {
    await new Promise(res =>
      createPages!(mockGatsbyContext, pluginOptions, res),
    )

    expect(mockGatsbyContext.actions.createNode).toMatchSnapshot()
  })

  test.skip('creates types', async () => {
    await new Promise(res =>
      createPages!(mockGatsbyContext, pluginOptions, res),
    )

    expect(mockGatsbyContext.actions.createTypes).toMatchSnapshot()
  })
})
