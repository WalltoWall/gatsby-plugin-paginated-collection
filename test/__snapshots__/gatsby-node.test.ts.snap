// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sourceNodes creates nodes 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "collectionName": "name",
        "hasNextPage": true,
        "hasPreviousPage": false,
        "id": "createNodeId",
        "index": 0,
        "internal": Object {
          "contentDigest": "createContentDigest",
          "type": "PaginatedCollectionPage",
        },
        "nextPage": "createNodeId",
        "nodeCount": 10,
        "nodes": Array [
          Object {
            "foo": 0,
          },
          Object {
            "foo": 1,
          },
          Object {
            "foo": 2,
          },
          Object {
            "foo": 3,
          },
          Object {
            "foo": 4,
          },
          Object {
            "foo": 5,
          },
          Object {
            "foo": 6,
          },
          Object {
            "foo": 7,
          },
          Object {
            "foo": 8,
          },
          Object {
            "foo": 9,
          },
        ],
        "previousPage": undefined,
      },
    ],
    Array [
      Object {
        "collectionName": "name",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "id": "createNodeId",
        "index": 1,
        "internal": Object {
          "contentDigest": "createContentDigest",
          "type": "PaginatedCollectionPage",
        },
        "nextPage": "createNodeId",
        "nodeCount": 10,
        "nodes": Array [
          Object {
            "foo": 10,
          },
          Object {
            "foo": 11,
          },
          Object {
            "foo": 12,
          },
          Object {
            "foo": 13,
          },
          Object {
            "foo": 14,
          },
          Object {
            "foo": 15,
          },
          Object {
            "foo": 16,
          },
          Object {
            "foo": 17,
          },
          Object {
            "foo": 18,
          },
          Object {
            "foo": 19,
          },
        ],
        "previousPage": "createNodeId",
      },
    ],
    Array [
      Object {
        "collectionName": "name",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "id": "createNodeId",
        "index": 2,
        "internal": Object {
          "contentDigest": "createContentDigest",
          "type": "PaginatedCollectionPage",
        },
        "nextPage": "createNodeId",
        "nodeCount": 10,
        "nodes": Array [
          Object {
            "foo": 20,
          },
          Object {
            "foo": 21,
          },
          Object {
            "foo": 22,
          },
          Object {
            "foo": 23,
          },
          Object {
            "foo": 24,
          },
          Object {
            "foo": 25,
          },
          Object {
            "foo": 26,
          },
          Object {
            "foo": 27,
          },
          Object {
            "foo": 28,
          },
          Object {
            "foo": 29,
          },
        ],
        "previousPage": "createNodeId",
      },
    ],
    Array [
      Object {
        "collectionName": "name",
        "hasNextPage": false,
        "hasPreviousPage": true,
        "id": "createNodeId",
        "index": 3,
        "internal": Object {
          "contentDigest": "createContentDigest",
          "type": "PaginatedCollectionPage",
        },
        "nextPage": undefined,
        "nodeCount": 5,
        "nodes": Array [
          Object {
            "foo": 30,
          },
          Object {
            "foo": 31,
          },
          Object {
            "foo": 32,
          },
          Object {
            "foo": 33,
          },
          Object {
            "foo": 34,
          },
        ],
        "previousPage": "createNodeId",
      },
    ],
    Array [
      Object {
        "firstPageSize": 10,
        "id": "createNodeId",
        "internal": Object {
          "contentDigest": "createContentDigest",
          "type": "PaginatedCollection",
        },
        "lastPageSize": 5,
        "name": "name",
        "nodeCount": 35,
        "pageCount": 4,
        "pageSize": 10,
        "pages": Array [
          "createNodeId",
          "createNodeId",
          "createNodeId",
          "createNodeId",
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`sourceNodes creates types 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "   \\"\\"\\"
  A page in a paginated collection.
  \\"\\"\\"
  type PaginatedCollectionPage implements Node @dontInfer {
    \\"Name of the collection to which this page belongs.\\"
    collectionName: String!

    \\"The next page.\\"
    nextPage: ID @link

    \\"Whether or not a page exists after this page.\\"
    hasNextPage: Boolean!

    \\"The previous page.\\"
    previousPage: ID @link

    \\"Whether or not a page exists before this page.\\"
    hasPreviousPage: Boolean!

    \\"Number of nodes in this page.\\"
    nodeCount: Int!

    \\"Nodes in this page.\\"
    nodes: [JSON!]!
  }

  \\"\\"\\"
  A paginated collection of nodes.
  \\"\\"\\"
  type PaginatedCollection implements Node @dontInfer {
    \\"Name of the collection.\\"
    name: String!

    \\"Maximum number of nodes in each page.\\"
    pageSize: Int!

    \\"Number of nodes in the first page.\\"
    firstPageSize: Int!

    \\"Number of nodes in the last page.\\"
    lastPageSize: Int!

    \\"Number of nodes in the collection.\\"
    nodeCount: Int!

    \\"Number of pages in the collection.\\"
    pageCount: Int!

    \\"Pages in the collection.\\"
    pages: [ID!]! @link
  }
",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
